---
title: "6.SCENIC"
author: "Jack Gordon"
date: "2024-04-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/jackgordon/Documents/DPhil/RNA-Sequencing/P10-seq")
```

Converts integrated seurat object to loom object for use in pySCENIC.\
pySCENIC was ran on HPC using command line interface.
Use SCENIC (R version) to identify differences in gene regulatory networks.

### 6. SCENIC
Convert integrated seurat object to loom for use in pySCENIC.
```{r, seurat to loom, message = FALSE, warning= FALSE, eval = FALSE}

#Load libraries
library(Seurat)
library(SeuratDisk)
library(SeuratObject)

#Import integrated seurat object, set assay to "RNA" and remove other assays
seurat <- readRDS("data/integrated.seurat.rds")
DefaultAssay(seurat) <- "RNA"

#Filter out genes that are expressed in less than 10 cells
seurat[["RNA"]] <- as(object = seurat[["RNA"]], Class = "Assay") #Convert RNA assay to v3
counts <- (seurat[["RNA"]]$counts)

# Output a logical vector for every gene on whether the more than zero counts per cell
nonzero <- counts > 0

# Sums all TRUE values and returns TRUE if more than 10 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 5

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]
filtered_counts <- Matrix::as.matrix(filtered_counts, sparse = TRUE)

# Reassign to filtered Seurat object
filtered_seurat <- CreateSeuratObject(filtered_counts, meta.data = seurat@meta.data)
filtered_seurat[["RNA"]] <- as(object = filtered_seurat[["RNA"]], Class = "Assay")

#Export seurat object to loom for use in pySCENIC
dir.create("pySCENIC/RNA/")
seurat.loom <- as.loom(x = filtered_seurat, filename = "pySCENIC/RNA/seurat.loom", verbose = TRUE)
```

pySCENIC CLI arguments (ran on HPC)
```{r pySCENIC cli, eval = FALSE, message = FALSE, warning = FALSE}
#GRN inference with GRNBoost2
!arboreto_with_multiprocessing.py  'RNA/seurat.loom' 'SCT/inputTFs.txt' -o adj.csv --num_workers 8 -m grnboost2

#Regulon prediction with cisTarget
!pyscenic ctx 'adj.csv' 'mm10__refseq-r80__10kb_up_and_down_tss.mc9nr.genes_vs_motifs.rankings.feather' 'mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.genes_vs_motifs.rankings.feather' --annotations_fname "motifs-v9-nr.mgi-m0.001-o0.0.tbl" --expression_mtx_fname 'seurat.loom' --output 'reg.csv' --mask_dropouts --num_workers 8

#Cellular enrichment with AUCell
!pyscenic aucell 'RNA/seurat.loom' 'RNA/reg.csv' --output 'RNA/pyscenic_output.loom' --num_workers 8

```

Downstream SCENIC analyses. 
```{r SCENIC, message = FALSE, warning = FALSE}
library(Seurat)
library(SeuratDisk)
library(SCENIC)
library(AUCell)
library(SCopeLoomR)
library(loomR)
library(ggplot2)
library(viridis)
library(dplyr)
library(ggpubr)
library(pheatmap)

#Read information from RNA loom file:
loom <- open_loom('pySCENIC/RNA/pyscenic_output.loom')
regulons_incidMat <- get_regulons(loom, column.attr.name = 'Regulons')
RNA_regulons <- regulonsToGeneLists(regulons_incidMat)
regulonsAUC <- get_regulons_AUC(loom, column.attr.name='RegulonsAUC')
AUCmat <- AUCell::getAUC(regulonsAUC)
rownames(AUCmat) <- gsub("[(+)]", "", rownames(AUCmat))

#Load seurat object, add AUCmat as a new assay and save updated object
seurat <- readRDS("data/integrated.seurat.rds")
seurat[["RNA_AUC"]] <- CreateAssayObject(data = AUCmat)
saveRDS(seurat, "data/integrated.seurat.rds")

#Read in seurat object with AUC matrix
seurat <- readRDS("data/integrated.seurat.rds")
seurat@active.ident <- factor(seurat$progenitor.origin)

#Plot heatmap for GRN across shared subtype clusters in integrated seurat
cellClusters <- seurat@meta.data

# Split the cells by integrated cluster to plot heatmap
cellsPerCluster <- split(rownames(cellClusters), cellClusters$integrated_snn_res.0.8)
regulonAUC <- regulonsAUC[onlyNonDuplicatedExtended(rownames(regulonsAUC)),]

# Calculate average expression:
regulonActivity_byCellType <- sapply(cellsPerCluster,
                                     function(cells) rowMeans(getAUC(regulonAUC)[,cells]))

# Scale expression:
regulonActivity_byCellType_Scaled <- t(scale(t(regulonActivity_byCellType), center = T, scale=T))

# Clean row names by removing '(+)'
cleaned_row_names <- gsub("\\(\\+\\)", "", rownames(regulonActivity_byCellType_Scaled))
rownames(regulonActivity_byCellType_Scaled) <- cleaned_row_names

# Plot heatmap
heatmap <- pheatmap(regulonActivity_byCellType_Scaled)

# Get the order of rows after clustering
row_order <- heatmap$tree_row$order

# Select labels for certain rows
selected_labels <- c("Dlx5", "Atf4", "Lhx8", "Lhx9", "Foxf2", "Neurod2", "Foxp2", 
                     "Otx2", "Pax6", "Setdb1", "E2f1", "Stat6", "Zfhx3", "Hdac2")

# Create a modified selected_row_names vector with empty space for non-displayed row names
selected_row_names <- ifelse(cleaned_row_names %in% selected_labels, cleaned_row_names, "")

# Draw the updated heatmap with modified row labels
heatmap <- pheatmap::pheatmap(
  regulonActivity_byCellType_Scaled,
  labels_row = selected_row_names,
  fontsize_row = 9  # Add legend title
)

#Find differentially active GRNs by progenitor origin

#Calculate average AUC for each grn by progenitor origin
cellClusters <- seurat@meta.data

# Split the cells by progenitor.origin
cellsPerCluster <- split(rownames(cellClusters), cellClusters$progenitor.origin)
regulonAUC <- regulonsAUC[onlyNonDuplicatedExtended(rownames(regulonsAUC)),]

# Calculate average expression and differences (raw and percent) for aIP and OP-derived cells
regulonActivity_byCellType <- sapply(cellsPerCluster,
                                     function(cells) rowMeans(getAUC(regulonAUC)[,cells]))

regulonActivity_byCellType <- data.frame(regulonActivity_byCellType)

rownames(regulonActivity_byCellType) <- gsub("[(+)]", "", rownames(regulonActivity_byCellType))

regulonActivity_byCellType$Difference <- regulonActivity_byCellType$aIP.derived - 
                                          regulonActivity_byCellType$OP.derived

regulonActivity_byCellType$percent.change <- ((regulonActivity_byCellType$aIP.derived - 
                                                 regulonActivity_byCellType$OP.derived) / 
                                                regulonActivity_byCellType$OP.derived) * 100

regulonActivity_byCellType$TF <- rownames(regulonActivity_byCellType)

#Find top 100 GRNs in aIP and OP-derived SPNs
aIP.grns <- regulonActivity_byCellType %>% arrange(desc(aIP.derived)) %>% slice_head(n = 100)
OP.grns <- regulonActivity_byCellType %>% arrange(desc(OP.derived)) %>% slice_head(n = 100)

shared_row_names <- intersect(rownames(aIP.grns), rownames(OP.grns)) #Shared GRNs
unique.aIP <- setdiff(rownames(aIP.grns), rownames(OP.grns)) #GRNs in aIP-derived top 100 but not OP-derived
unique.OP <- setdiff(rownames(OP.grns), rownames(aIP.grns)) ##GRNs in OP-derived top 100 but not aIP-derived

selected_TFs <- c(shared_row_names, unique.aIP, unique.OP) #Selected_TFs for downstream analyses

#Perform DE testing - to filter for signifcant GRNs based on adjusted p_values
seurat@active.ident <- as.factor(seurat$progenitor.origin)

grns <- FindMarkers(seurat, ident.1 = "aIP-derived", 
                    ident.2 = "OP-derived", assay = "RNA_AUC", 
                    logfc.threshold = 0) #No threshold as AUC score is enrichment

grns <- grns %>% filter(p_val_adj < 0.05) #Keep significantly DE GRNs

grns.to.plot <- regulonActivity_byCellType %>% filter(TF %in% selected_TFs) %>%
  filter(TF %in% rownames(grns)) %>%
  mutate(progenitor.bias = ifelse(percent.change > 0, 
                                  "aIP-derived", "OP-derived")) %>%
  arrange(desc(TF)) #Create dataeframe of significant GRNs to plot

#Plot percent.change for the DE GRNs
ggbarplot(grns.to.plot, x = "TF", y = "percent.change", 
          fill = "progenitor.bias", color = "black", 
          palette = c("forestgreen", "firebrick3"),
          xlab = "Gene Regulatory Network (GRN)", ylab = "Difference in GRN Activity (%)") + 
  theme_classic(base_size = 15) + 
  theme(legend.position = "none") +
  coord_flip()

```

```{r, network graph, eval = FALSE, message = FALSE, warning = FALSE}
#Create network graph of GRNs in aIP and OP-derived cells
library(igraph)

regulons <- RNA_regulons #Load regulons (TFs and regulated genes)

#Create an empty edgelist
edgelist <- matrix(ncol = 2, nrow = 0, dimnames = list(NULL, c("TF", "Target")))

#Populate the edgelist with TF-gene connections
for (tf in names(regulons)) {
  tf_name <- gsub("\\(\\+\\)", "", tf)  # Remove '(+)' from TF names
  if (tf_name %in% selected_TFs) {
    for (gene in regulons[[tf]]) {
      edgelist <- rbind(edgelist, c(tf_name, gene))
    }
  }
}

#Create a directed graph from the edgelist
g <- graph_from_edgelist(as.matrix(edgelist), directed = TRUE)

#Find TFs that are regulated by other TFs
regulated_TFs <- V(g)$name[degree(g, mode = "in") > 0 & degree(g, mode = "out") > 0]

#Filter the edgelist to include only regulated TFs and their connections to other TFs
filtered_edgelist <- edgelist[edgelist[,1] %in% regulated_TFs & edgelist[,2] %in% regulated_TFs, ]

# Create a new graph from the filtered edgelist
subgraph <- graph_from_edgelist(as.matrix(filtered_edgelist), directed = TRUE, )

# Get the induced subgraph of the selected TFs and their connections
selected_TF_names <- names(V(subgraph))  # Get names of vertices in the subgraph
selected_TF_indices <- which(selected_TF_names %in% selected_TFs)  # Find indices of selected TFs
subgraph <- induced_subgraph(subgraph, vids = selected_TF_indices)

#Plot using tkplot - adjust manually to colour differentially active GRNs
tkplot(subgraph, vertex.label.cex = 2, label.font = "Arial", label.color = "black", edge.width = 0.5, 
       arrow.size = 0.5, vertex.size = 20)
tk_off()
```


![](scenic.tfs.pdf){width=800px} 


```{r, RSS, warning = FALSE, message = FALSE}
# Calculate regulon specificity scores
AUC <- getAUC(regulonsAUC)

#Clean AUC regulon_names
current_regulon_names <- dimnames(AUC)[[1]]
cleaned_regulon_names <- sub("\\(\\+\\)$", "", current_regulon_names)
dimnames(AUC)[[1]] <- cleaned_regulon_names

#Filter for regulons of interest
regulon_names <- rownames(AUC)

# Identify indices of regulons that belong to selected transcription factors
selected_indices <- which(grepl(paste(selected_TFs, collapse = "|"), regulon_names))

# Filter the AUC matrix to include only regulons from selected transcription factors
filtered_AUC <- AUC[selected_indices, , drop = FALSE]


rss <- calcRSS(filtered_AUC, cellAnnotation=cellClusters[colnames(regulonsAUC), 
                                                                    "progenitor.origin"])

plotRSS_oneSet(rss, setName = "aIP-derived")
plotRSS_oneSet(rss, setName = "OP-derived")
```

```{r sessionInfo, message = FALSE, warning = FALSE}
sessionInfo()
```

